File: mini-jdbc/src/main/java/com/mini/core/jdbc/builder/SqlBuilder.java
package com.mini.core.jdbc.builder;

import org.springframework.util.Assert;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

import static com.mini.core.jdbc.builder.SQLInterface.sqlInterface;
import static java.lang.String.format;
import static java.lang.String.join;
import static java.util.Arrays.stream;
import static java.util.Collections.addAll;
import static java.util.stream.Collectors.joining;

/**
 * SQL构建器
 *
 * @author xchao
 */
@SuppressWarnings("UnusedReturnValue")
public class SQLBuilder implements Serializable {
    private final OnDuplicateKeyUpdateStatement ON_DUPLICATE_KEY_UPDATE = new OnDuplicateKeyUpdateStatement();
    private final OuterJoinStatement OUTER_JOIN = new OuterJoinStatement();
    private final InnerJoinStatement INNER_JOIN = new InnerJoinStatement();
    private final RightJoinStatement RIGHT_JOIN = new RightJoinStatement();
    private final LeftJoinStatement LEFT_JOIN = new LeftJoinStatement();
    private final GroupByStatement GROUP_BY = new GroupByStatement();
    private final OrderByStatement ORDER_BY = new OrderByStatement();
    private final ColumnStatement COLUMNS = new ColumnStatement();
    private final HavingStatement HAVING = new HavingStatement();
    private final ValuesStatement VALUES = new ValuesStatement();
    private final SelectStatement SELECT = new SelectStatement();
    private final TableStatement TABLE = new TableStatement();
    private final WhereStatement WHERE = new WhereStatement();
    private final FromStatement FROM = new FromStatement();
    private final JoinStatement JOIN = new JoinStatement();
    private final SetStatement SET = new SetStatement();
    private StatementType STATEMENT = null;
    private WhereStatement LAST = null;
    private boolean DISTINCT = false;
    private final List<Object> args = new ArrayList<>();

    public SQLBuilder() {
    }

    public SQLBuilder(Consumer<SQLBuilder> consumer) {
        consumer.accept(this);
    }

    public final SQLBuilder ARGS(Object... args) {
        addAll(this.args, args);
        return this;
    }

    public final Object[] getArgs() {
        return args.toArray();
    }

    public final SQLBuilder INSERT_INTO(String table) {
        STATEMENT = StatementType.INSERT;
        TABLE.addValues(table);
        return this;
    }

    public final <T> SQLBuilder INSERT_INTO(T instance) {
        sqlInterface.createInsert(this, instance);
        return this;
    }

    public final SQLBuilder REPLACE_INTO(String table) {
        STATEMENT = StatementType.REPLACE;
        TABLE.addValues(table);
        return this;
    }

    public final <T> SQLBuilder REPLACE_INTO(T instance) {
        sqlInterface.createReplace(this, instance);
        return this;
    }

    public final SQLBuilder DELETE(String... tables) {
        STATEMENT = StatementType.DELETE;
        TABLE.addValues(tables);
        return this;
    }

    public final <T> SQLBuilder DELETE(T instance) {
        sqlInterface.createDelete(this, instance);
        return this;
    }

    public final SQLBuilder UPDATE(String... tables) {
        STATEMENT = StatementType.UPDATE;
        TABLE.addValues(tables);
        return this;
    }

    public final <T> SQLBuilder UPDATE(T instance) {
        sqlInterface.createUpdate(this, instance);
        return this;
    }

    public final <T> SQLBuilder SELECT_FROM_JOIN(Class<T> type) {
        sqlInterface.createSelect(this, type);
        return this;
    }

    public final SQLBuilder SELECT(String... columns) {
        STATEMENT = StatementType.SELECT;
        SELECT.addSelect(columns);
        return this;
    }

    public final SQLBuilder SELECT_DISTINCT(String... columns) {
        SQLBuilder.this.DISTINCT = Boolean.TRUE;
        SQLBuilder.this.SELECT(columns);
        return this;
    }

    public final SQLBuilder VALUES(String column, String value) {
        COLUMNS.addValues(column);
        VALUES.addValues(value);
        return this;
    }

    public final SQLBuilder VALUES(String column) {
        return VALUES(column, "?");
    }


    public final SQLBuilder FROM(String... tables) {
        FROM.addValues(tables);
        return this;
    }

    public final SQLBuilder FROM(SQLBuilder builder) {
        return FROM(builder.toString());
    }

    public final SQLBuilder JOIN(String join) {
        JOIN.addValues(join);
        return this;
    }

    public final SQLBuilder INNER_JOIN(String join) {
        INNER_JOIN.addValues(join);
        return this;
    }

    public final SQLBuilder LEFT_JOIN(String join) {
        LEFT_JOIN.addValues(join);
        return this;
    }

    public final SQLBuilder RIGHT_JOIN(String join) {
        RIGHT_JOIN.addValues(join);
        return this;
    }

    public final SQLBuilder OUTER_JOIN(String join) {
        OUTER_JOIN.addValues(join);
        return this;
    }

    public final SQLBuilder SET(String set) {
        SET.addValues(set);
        return this;
    }

    public final SQLBuilder SET_EQUALS(String column, @Nullable Object arg) {
        return SET(column + " = ?").ARGS(arg);
    }

    public final SQLBuilder SET_INCREASE(String column, @Nullable Object arg) {
        return SET(column + " = " + column + " + ?").ARGS(arg);
    }

    public final <T> SQLBuilder ON_DUPLICATE_KEY_UPDATE(T instance) {
        sqlInterface.createInsertOnUpdate(this, instance);
        return this;
    }

    public final SQLBuilder ON_DUPLICATE_KEY_UPDATE(String set) {
        this.STATEMENT = StatementType.INSERT_UPDATE;
        ON_DUPLICATE_KEY_UPDATE.addValues(set);
        return this;
    }

    public final SQLBuilder ON_DUPLICATE_KEY_UPDATE_FROM_INSERT(String column) {
        return ON_DUPLICATE_KEY_UPDATE(column + " VALUES(" + column + ")");
    }

    public final SQLBuilder ON_DUPLICATE_KEY_UPDATE_EQUALS(String column, @Nullable Object arg) {
        return ON_DUPLICATE_KEY_UPDATE(column + " = ?").ARGS(arg);
    }

    public final SQLBuilder ON_DUPLICATE_KEY_UPDATE_INCREASE(String column, @Nullable Object arg) {
        return ON_DUPLICATE_KEY_UPDATE(column + " = " + column + " + ?").ARGS(arg);
    }


    public final SQLBuilder AND() {
        if (this.LAST != null) {
            LAST.addAND();
        }
        return this;
    }

    public final SQLBuilder OR() {
        if (this.LAST != null) {
            LAST.addOR();
        }
        return this;
    }

    public final SQLBuilder WHERE(String where) {
        this.WHERE.addValues(where);
        this.LAST = this.WHERE;
        return this;
    }

    public final SQLBuilder WHERE_IN(String column, Object... args) {
        Assert.notEmpty(args, "WHERE_IN args cannot empty");
        WHERE(format("%s IN(%s)", column, stream(args).map(it -> "?").collect(joining(", "))));
        return ARGS(args);
    }

    public final SQLBuilder WHERE_EQUALS(String column, @Nullable Object arg) {
        if (arg == null) return WHERE(column + " IS NULL");
        return WHERE(column + " = ?").ARGS(arg);
    }

    public final SQLBuilder WHERE_NOT_EQUALS(String column, @Nullable Object arg) {
        if (arg == null) return WHERE(column + " IS NOT NULL");
        return WHERE(column + " <> ?").ARGS(arg);
    }

    public final SQLBuilder WHERE_MATCH(String[] columns, String args) {
        Assert.notEmpty(columns, "WHERE_MATCH args cannot empty");
        return WHERE("MATCH(" + join(", ", columns) + ") AGAINST(?)").ARGS(args);
    }

    public final SQLBuilder WHERE_MATCH_IN_BOOL(String[] columns, String args) {
        Assert.notEmpty(columns, "WHERE_MATCH args cannot empty");
        return WHERE("MATCH(" + join(", ", columns) + ") AGAINST(? in BOOLEAN MODE)").ARGS(args);
    }

    public final SQLBuilder GROUP_BY(String... columns) {
        GROUP_BY.addValues(columns);
        return this;
    }

    public final SQLBuilder HAVING(String having) {
        this.HAVING.addValues(having);
        this.LAST = this.HAVING;
        return this;
    }

    public final SQLBuilder HAVING_IN(String column, Object... args) {
        Assert.notEmpty(args, "HAVING_IN args cannot empty");
        HAVING(format("%s IN(%s)", column, stream(args).map(it -> "?").collect(joining(", "))));
        return ARGS(args);
    }

    public final SQLBuilder HAVING_EQUALS(String column, @Nullable Object arg) {
        if (arg == null) return HAVING(column + " IS NULL");
        return HAVING(column + " = ?").ARGS(arg);
    }

    public final SQLBuilder HAVING_NOT_EQUALS(String column, @Nullable Object arg) {
        if (arg == null) return HAVING(column + " IS NOT NULL");
        return HAVING(column + " <> ?").ARGS(arg);
    }

    public final SQLBuilder HAVING_MATCH(String[] columns, String args) {
        Assert.notEmpty(columns, "WHERE_MATCH args cannot empty");
        return HAVING("MATCH(" + join(", ", columns) + ") AGAINST(?)").ARGS(args);
    }

    public final SQLBuilder HAVING_MATCH_IN_BOOL(String[] columns, String args) {
        Assert.notEmpty(columns, "WHERE_MATCH args cannot empty");
        return HAVING("MATCH(" + join(", ", columns) + ") AGAINST(? in BOOLEAN MODE)").ARGS(args);
    }

    public final SQLBuilder ORDER_BY(String orderBy) {
        ORDER_BY.addValues(orderBy);
        return this;
    }

    // Insert Into
    private String INSERT_STRING() {
        var builder = new StringBuilder();
        builder.append("INSERT INTO ");
        TABLE.builder(builder);
        COLUMNS.builder(builder);
        VALUES.builder(builder);
        return builder.toString();
    }

    // Replace Into
    private String REPLACE_STRING() {
        var builder = new StringBuilder();
        builder.append("REPLACE INTO ");
        TABLE.builder(builder);
        COLUMNS.builder(builder);
        VALUES.builder(builder);
        return builder.toString();
    }

    // Delete Into
    private String DELETE_STRING() {
        var builder = new StringBuilder();
        TABLE.builder(builder.append("DELETE "));
        FROM.builder(builder);
        JOIN.builder(builder);
        LEFT_JOIN.builder(builder);
        RIGHT_JOIN.builder(builder);
        OUTER_JOIN.builder(builder);
        WHERE.builder(builder);
        return builder.toString();
    }

    // Update Into
    private String UPDATE_STRING() {
        var builder = new StringBuilder();
        TABLE.builder(builder.append("UPDATE "));
        JOIN.builder(builder);
        LEFT_JOIN.builder(builder);
        RIGHT_JOIN.builder(builder);
        OUTER_JOIN.builder(builder);
        SET.builder(builder);
        WHERE.builder(builder);
        return builder.toString();
    }

    // Select Into
    private String SELECT_STRING() {
        var builder = new StringBuilder("SELECT ");
        if (SQLBuilder.this.DISTINCT) {
            builder.append("DISTINCT ");
        }
        SELECT.builder(builder);
        FROM.builder(builder);
        JOIN.builder(builder);
        LEFT_JOIN.builder(builder);
        RIGHT_JOIN.builder(builder);
        OUTER_JOIN.builder(builder);
        WHERE.builder(builder);
        GROUP_BY.builder(builder);
        HAVING.builder(builder);
        ORDER_BY.builder(builder);
        return builder.toString();
    }

    private String INSERT_ON_UPDATE_STRING() {
        var builder = new StringBuilder();
        builder.append("INSERT INTO ");
        TABLE.builder(builder);
        COLUMNS.builder(builder);
        VALUES.builder(builder);
        ON_DUPLICATE_KEY_UPDATE.builder(builder);
        return builder.toString();
    }

    public final String getSql() {
        //  INSERT 语句
        if (STATEMENT == StatementType.INSERT) {
            return INSERT_STRING();
        }
        // REPLACE 语句
        if (STATEMENT == StatementType.REPLACE) {
            return REPLACE_STRING();
        }
        // DELETE 语句
        if (STATEMENT == StatementType.DELETE) {
            return DELETE_STRING();
        }
        // UPDATE 语句
        if (STATEMENT == StatementType.UPDATE) {
            return UPDATE_STRING();
        }
        // SELECT 语句
        if (STATEMENT == StatementType.SELECT) {
            return SELECT_STRING();
        }
        // INSERT INTO ON DUPLICATE KEY UPDATE
        if (STATEMENT == StatementType.INSERT_UPDATE) {
            return INSERT_ON_UPDATE_STRING();
        }
        throw new RuntimeException("SQL ERROR!");
    }

    @Override
    public final String toString() {
        return getSql() + "\n" + Arrays.toString(getArgs());
    }

    // Base Statement
    private static abstract class BaseStatement implements Serializable {
        protected final List<String> values = new ArrayList<>();
        public static final String AND = ") AND (";
        public static final String OR = ") OR (";
        final String keyWord, join;

        public BaseStatement(String keyWord, String join) {
            this.keyWord = keyWord;
            this.join = join;
        }

        @Nonnull
        public String getOpen() {
            return "";
        }

        @Nonnull
        public String getClose() {
            return "";
        }

        public final void addValues(String... values) {
            if (values != null && values.length > 0) {
                addAll(this.values, values);
            }
        }

        public final void builder(StringBuilder builder) {
            if (BaseStatement.this.values.isEmpty()) {
                return;
            }
            builder.append(keyWord).append(getOpen());
            String last = "_________________________";
            for (int i = 0; i < values.size(); i++) {
                String part = this.values.get(i);
                if (this.isJoin(i, part, last)) {
                    builder.append(join);
                }
                builder.append(part);
                last = part;
            }
            builder.append(getClose());
        }

        private boolean isJoin(int index, String part, String last) {
            var b = index > 0 && !OR.equals(part) && !OR.equals(last);
            return b && !AND.equals(part) && !AND.equals(last);
        }
    }

    // Table Statement
    private static class TableStatement extends BaseStatement {
        public TableStatement() {
            super("", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }
    }

    // Column Statement
    private static class ColumnStatement extends BaseStatement {
        public ColumnStatement() {
            super("", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "(";
        }

        @Nonnull
        public final String getClose() {
            return ") ";
        }
    }

    // Values Statement
    private static class ValuesStatement extends BaseStatement {
        public ValuesStatement() {
            super("\nVALUES ", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "(";
        }

        @Nonnull
        public final String getClose() {
            return ")";
        }
    }

    // Field Statement
    private static class SelectStatement extends BaseStatement {
        public SelectStatement() {
            super("", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }

        public final void addSelect(String... columns) {
            if (columns != null && columns.length > 0) {
                columns[0] = "\n\t" + columns[0];
                super.addValues(columns);
            }
        }
    }

    // From Statement
    private static class FromStatement extends BaseStatement {
        public FromStatement() {
            super("\nFROM ", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }
    }

    // Join Statement
    private static class JoinStatement extends BaseStatement {

        public JoinStatement(String word) {
            super(word, word);
        }

        public JoinStatement() {
            this("\nJOIN ");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }

    }

    // inner Join Statement
    private static class InnerJoinStatement extends JoinStatement {
        public InnerJoinStatement() {
            super("\nINNER JOIN ");
        }
    }

    // Left Join Statement
    private static class LeftJoinStatement extends JoinStatement {
        public LeftJoinStatement() {
            super("\nLEFT JOIN ");
        }
    }

    // Right Join Statement
    private static class RightJoinStatement extends JoinStatement {
        public RightJoinStatement() {
            super("\nRIGHT JOIN ");
        }
    }

    // Outer Join Statement
    private static class OuterJoinStatement extends JoinStatement {
        public OuterJoinStatement() {
            super("\nOUTER JOIN ");
        }
    }

    // Set Statement
    private static class SetStatement extends BaseStatement {
        public SetStatement() {
            super("\nSET ", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }
    }

    private static class OnDuplicateKeyUpdateStatement extends BaseStatement {
        public OnDuplicateKeyUpdateStatement() {
            super("\nON DUPLICATE KEY UPDATE ", ",");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }
    }

    // Where Statement
    private static class WhereStatement extends BaseStatement {
        public WhereStatement(String word) {
            super(word, AND);
        }

        public WhereStatement() {
            this("\nWHERE ");
        }

        @Nonnull
        public final String getOpen() {
            return "(";
        }

        @Nonnull
        public final String getClose() {
            return ") ";
        }

        public final void addAND() {
            values.add(AND);
        }

        public final void addOR() {
            values.add(OR);
        }
    }

    // Group By Statement
    private static class GroupByStatement extends BaseStatement {
        public GroupByStatement() {
            super("\nGROUP BY ", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }
    }

    private static class HavingStatement extends WhereStatement {
        public HavingStatement() {
            super("\nHAVING ");
        }
    }

    // Order By Statement
    private static class OrderByStatement extends BaseStatement {
        public OrderByStatement() {
            super("\nORDER BY ", ", ");
        }

        @Nonnull
        public final String getOpen() {
            return "";
        }

        @Nonnull
        public final String getClose() {
            return " ";
        }
    }

    private enum StatementType {
        INSERT, REPLACE, DELETE, UPDATE, SELECT, INSERT_UPDATE
    }
}

File: mini-jdbc/src/main/kotlin/com/mini/core/jdbc/builder/SQLBuilder.kt
@file:Suppress("FunctionName", "PrivatePropertyName", "MemberVisibilityCanBePrivate", "unused", "PropertyName")

package com.mini.core.jdbc.builder

import com.mini.core.jdbc.builder.SQLInterfaceDef.Companion.getSQLInterface
import java.io.Serializable
import java.util.*
import java.util.function.Consumer
import kotlin.reflect.KClass

/**
 * SQL构建器
 *
 * 构建结果会根据最后一次提交更改的类型构建，其它与该构建类型不相关的子句将会被忽略
 *
 *
 *
 * @author xchao
 */
class SQLBuilder constructor(init: SQLBuilder.() -> Unit) : EventListener, Serializable {
    private val ON_DUPLICATE_KEY_UPDATE = OnDuplicateKeyUpdateStatement()
    private val OUTER_JOIN = OuterJoinStatement()
    private val INNER_JOIN = InnerJoinStatement()
    private val RIGHT_JOIN = RightJoinStatement()
    private var STATEMENT: StatementType? = null
    private val LEFT_JOIN = LeftJoinStatement()
    private val GROUP_BY = GroupByStatement()
    private val ORDER_BY = OrderByStatement()
    private var LAST: WhereStatement? = null
    private val COLUMNS = ColumnStatement()
    private val HAVING = HavingStatement()
    private val VALUES = ValuesStatement()
    private val SELECT = SelectStatement()
    private val TABLE = TableStatement()
    private val WHERE = WhereStatement()
    private val FROM = FromStatement()
    private val JOIN = JoinStatement()
    private val SET = SetStatement()
    private var DISTINCT = false

    val ARGS: MutableList<Any?> = mutableListOf()
    val SQL: String
        get() {
            //  INSERT 语句
            if (STATEMENT == StatementType.INSERT) {
                return INSERT_STRING()
            }
            // REPLACE 语句
            if (STATEMENT == StatementType.REPLACE) {
                return REPLACE_STRING()
            }
            // DELETE 语句
            if (STATEMENT == StatementType.DELETE) {
                return DELETE_STRING()
            }
            // UPDATE 语句
            if (STATEMENT == StatementType.UPDATE) {
                return UPDATE_STRING()
            }
            // SELECT 语句
            if (STATEMENT == StatementType.SELECT) {
                return SELECT_STRING()
            }
            // INSERT INTO ON DUPLICATE KEY UPDATE
            if (STATEMENT == StatementType.INSERT_UPDATE) {
                return INSERT_ON_UPDATE_STRING()
            }
            throw RuntimeException("SQL ERROR!")
        }

    init {
        this.apply(init)
    }

    constructor(consumer: Consumer<SQLBuilder>) : this({
        consumer.accept(this)
    })

    fun ARGS(vararg args: Any?): SQLBuilder {
        ARGS.addAll(listOf(args))
        return this
    }

    fun INSERT_INTO(table: String): SQLBuilder {
        STATEMENT = StatementType.INSERT
        TABLE.ADD_VALUES(table)
        return this
    }

    fun <T : Any> INSERT_INTO(instance: T): SQLBuilder {
        getSQLInterface(instance::class as KClass<*>).createInsert(this, instance)
        return this
    }

    fun REPLACE_INTO(table: String): SQLBuilder {
        STATEMENT = StatementType.REPLACE
        TABLE.ADD_VALUES(table)
        return this
    }

    fun <T : Any> REPLACE_INTO(instance: T): SQLBuilder {
        getSQLInterface(instance::class as KClass<*>).createReplace(this, instance)
        return this
    }

    fun DELETE(vararg tables: String): SQLBuilder {
        STATEMENT = StatementType.DELETE
        TABLE.ADD_VALUES(*tables)
        return this
    }

    fun UPDATE(vararg tables: String): SQLBuilder {
        STATEMENT = StatementType.UPDATE
        TABLE.ADD_VALUES(*tables)
        return this
    }

    fun <T : Any> UPDATE(instance: T): SQLBuilder {
        getSQLInterface(instance::class as KClass<*>).createUpdate(this, instance)
        return this
    }

    fun <T : Any> SELECT_FROM_JOIN(type: KClass<T>): SQLBuilder {
        getSQLInterface(type).createSelect(this, type)
        return this
    }

    fun SELECT(vararg columns: String): SQLBuilder {
        STATEMENT = StatementType.SELECT
        SELECT.ADD_SELECT(*columns)
        return this
    }

    fun SELECT_DISTINCT(vararg columns: String): SQLBuilder {
        DISTINCT = true
        SELECT(*columns)
        return this
    }

    @JvmOverloads
    fun VALUES(column: String, value: String = "?"): SQLBuilder {
        COLUMNS.ADD_VALUES(column)
        VALUES.ADD_VALUES(value)
        return this
    }

    fun FROM(vararg tables: String): SQLBuilder {
        FROM.ADD_VALUES(*tables)
        return this
    }

    fun FROM(builder: SQLBuilder): SQLBuilder {
        return FROM(builder.toString())
    }

    fun JOIN(join: String): SQLBuilder {
        JOIN.ADD_VALUES(join)
        return this
    }

    fun INNER_JOIN(join: String): SQLBuilder {
        INNER_JOIN.ADD_VALUES(join)
        return this
    }

    fun LEFT_JOIN(join: String): SQLBuilder {
        LEFT_JOIN.ADD_VALUES(join)
        return this
    }

    fun RIGHT_JOIN(join: String): SQLBuilder {
        RIGHT_JOIN.ADD_VALUES(join)
        return this
    }

    fun OUTER_JOIN(joins: String): SQLBuilder {
        OUTER_JOIN.ADD_VALUES(joins)
        return this
    }

    fun SET(sets: String): SQLBuilder {
        SET.ADD_VALUES(sets)
        return this
    }

    fun ON_DUPLICATE_KEY_UPDATE(sets: String): SQLBuilder {
        ON_DUPLICATE_KEY_UPDATE.ADD_VALUES(sets)
        STATEMENT = StatementType.INSERT_UPDATE
        return this
    }

    fun <T : Any> ON_DUPLICATE_KEY_UPDATE(instance: T): SQLBuilder {
        getSQLInterface(instance::class as KClass<*>).createInsertOnUpdate(this, instance)
        return this
    }

    fun AND(): SQLBuilder {
        if (LAST != null) {
            LAST!!.AND()
        }
        return this
    }

    fun OR(): SQLBuilder {
        if (LAST != null) {
            LAST!!.OR()
        }
        return this
    }

    fun WHERE(wheres: String): SQLBuilder {
        WHERE.ADD_VALUES(wheres)
        LAST = WHERE
        return this
    }

    fun WHERE_IN(column: String, args: Array<Any?>): SQLBuilder {
        return WHERE("""$column IN (${args.joinToString { "?" }}""").ARGS(*args)
    }

    fun WHERE_MATCH(vararg column: String, args: String): SQLBuilder {
        return WHERE("MATCH(${column.joinToString()}) AGAINST(?)").ARGS(args)
    }

    fun WHERE_MATCH_IN_BOOL(vararg column: String, args: String): SQLBuilder {
        return WHERE("MATCH(${column.joinToString()}) AGAINST(?)").ARGS(args)
    }

    fun GROUP_BY(vararg columns: String): SQLBuilder {
        GROUP_BY.ADD_VALUES(*columns)
        return this
    }

    fun HAVING(having: String): SQLBuilder {
        HAVING.ADD_VALUES(having)
        LAST = HAVING
        return this
    }

    fun HAVING_IN(column: String, args: Array<Any?>): SQLBuilder {
        return HAVING("""$column IN (${args.joinToString(", ") { "?" }}""").ARGS(*args)
    }

    fun HAVING_MATCH(vararg column: String, args: String): SQLBuilder {
        return HAVING("MATCH(${column.joinToString()}) AGAINST(?)").ARGS(args)
    }

    fun HAVING_MATCH_IN_BOOL(vararg column: String, args: String): SQLBuilder {
        return HAVING("MATCH(${column.joinToString()}) AGAINST(?)").ARGS(args)
    }

    fun ORDER_BY(orderBy: String): SQLBuilder {
        ORDER_BY.ADD_VALUES(orderBy)
        return this
    }

    // Insert Into
    private fun INSERT_STRING(): String {
        val builder = StringBuilder()
        builder.append("INSERT INTO ")
        TABLE.BUILDER(builder)
        COLUMNS.BUILDER(builder)
        VALUES.BUILDER(builder)
        return builder.toString()
    }

    // Replace Into
    private fun REPLACE_STRING(): String {
        val builder = StringBuilder()
        builder.append("REPLACE INTO ")
        TABLE.BUILDER(builder)
        COLUMNS.BUILDER(builder)
        VALUES.BUILDER(builder)
        return builder.toString()
    }

    // Delete Into
    private fun DELETE_STRING(): String {
        val builder = StringBuilder()
        TABLE.BUILDER(builder.append("DELETE "))
        FROM.BUILDER(builder)
        JOIN.BUILDER(builder)
        LEFT_JOIN.BUILDER(builder)
        RIGHT_JOIN.BUILDER(builder)
        OUTER_JOIN.BUILDER(builder)
        WHERE.BUILDER(builder)
        return builder.toString()
    }

    // Update Into
    private fun UPDATE_STRING(): String {
        val builder = StringBuilder()
        TABLE.BUILDER(builder.append("UPDATE "))
        JOIN.BUILDER(builder)
        LEFT_JOIN.BUILDER(builder)
        RIGHT_JOIN.BUILDER(builder)
        OUTER_JOIN.BUILDER(builder)
        SET.BUILDER(builder)
        WHERE.BUILDER(builder)
        return builder.toString()
    }

    // Select Into
    private fun SELECT_STRING(): String {
        val builder = StringBuilder("SELECT ")
        builder.append(if (DISTINCT) "DISTINCT " else "")
        SELECT.BUILDER(builder)
        FROM.BUILDER(builder)
        JOIN.BUILDER(builder)
        LEFT_JOIN.BUILDER(builder)
        RIGHT_JOIN.BUILDER(builder)
        OUTER_JOIN.BUILDER(builder)
        WHERE.BUILDER(builder)
        GROUP_BY.BUILDER(builder)
        HAVING.BUILDER(builder)
        ORDER_BY.BUILDER(builder)
        return builder.toString()
    }

    private fun INSERT_ON_UPDATE_STRING(): String {
        val builder = StringBuilder()
        builder.append("INSERT INTO ")
        TABLE.BUILDER(builder)
        COLUMNS.BUILDER(builder)
        VALUES.BUILDER(builder)
        ON_DUPLICATE_KEY_UPDATE.BUILDER(builder)
        return builder.toString()
    }

    @Synchronized
    override fun toString(): String {
        return "${SQL}\n$ARGS"
    }

    // Base Statement
    private abstract class BaseStatement constructor(val keyWord: String, val join: String) {
        protected val values: MutableList<String> = mutableListOf()

        companion object {
            const val AND = ") AND ("
            const val OR = ") OR ("
        }

        protected open fun OPEN(): String {
            return ""
        }

        protected open fun CLOSE(): String {
            return ""
        }

        fun ADD_VALUES(vararg values: String) {
            values.takeIf { it.isNotEmpty() }?.let {
                this.values.addAll(values)
            }
        }

        fun BUILDER(builder: StringBuilder) {
            this.values.takeIf { it.isNotEmpty() }?.let {
                builder.append(keyWord).append(OPEN())
                var last = "_________________________"
                for (i in values.indices) {
                    val part = values[i]
                    if (IS_JOIN(i, part, last)) {
                        builder.append(join)
                    }
                    builder.append(part)
                    last = part
                }
                builder.append(CLOSE())
            }
        }

        private fun IS_JOIN(index: Int, part: String, last: String): Boolean {
            return index > 0 && AND != part && AND != last && OR != part && OR != last
        }
    }

    // Table Statement
    private class TableStatement : BaseStatement("", ", ") {
        override fun OPEN(): String {
            return ""
        }

        override fun CLOSE(): String {
            return " "
        }
    }

    // Column Statement
    private class ColumnStatement : BaseStatement("", ", ") {
        override fun OPEN(): String {
            return "("
        }

        override fun CLOSE(): String {
            return ") "
        }
    }

    // Values Statement
    private class ValuesStatement : BaseStatement("\nVALUES ", ", ") {
        override fun OPEN(): String {
            return "("
        }

        override fun CLOSE(): String {
            return ")"
        }
    }

    // Field Statement
    private class SelectStatement : BaseStatement("", ", ") {
        override fun OPEN(): String {
            return ""
        }

        override fun CLOSE(): String {
            return " "
        }

        fun ADD_SELECT(vararg columns: String) = mutableListOf(*columns).takeIf {
            it.isNotEmpty()
        }?.let {
            it.set(index = 0, element = "\t\n${it[0]}")
            this@SelectStatement.values.addAll(it)
        }
    }

    // From Statement
    private class FromStatement : BaseStatement("\nFROM ", ", ") {
        override fun OPEN(): String {
            return ""
        }

        override fun CLOSE(): String {
            return " "
        }
    }

    // Join Statement
    private open class JoinStatement constructor(word: String = "\nJOIN ") : BaseStatement(word, word) {

        override fun OPEN(): String {
            return ""
        }


        override fun CLOSE(): String {
            return " "
        }
    }

    // Inner Join Statement
    private class InnerJoinStatement constructor(word: String = "\nINNER JOIN ") : JoinStatement(word)

    // Left Join Statement
    private class LeftJoinStatement constructor(word: String = "\nLEFT JOIN ") : JoinStatement(word)

    // Right Join Statement
    private class RightJoinStatement constructor(word: String = "\nRIGHT JOIN ") : JoinStatement(word)

    // Outer Join Statement
    private class OuterJoinStatement constructor(word: String = "\nOUTER JOIN ") : JoinStatement(word)

    // Set Statement
    private class SetStatement : BaseStatement("\nSET ", ", ") {
        override fun OPEN(): String {
            return ""
        }

        override fun CLOSE(): String {
            return " "
        }
    }

    private class OnDuplicateKeyUpdateStatement : BaseStatement("\nON DUPLICATE KEY UPDATE ", ",") {
        override fun OPEN(): String {
            return ""
        }

        override fun CLOSE(): String {
            return " "
        }
    }

    // Where Statement
    private open class WhereStatement constructor(word: String = "\nWHERE ") : BaseStatement(word, AND) {
        override fun OPEN(): String {
            return "("
        }

        override fun CLOSE(): String {
            return ") "
        }

        fun AND() {
            values.add(AND)
        }

        fun OR() {
            values.add(OR)
        }
    }

    // Group By Statement
    private class GroupByStatement : BaseStatement("\nGROUP BY ", ", ") {
        override fun OPEN(): String {
            return ""
        }

        override fun CLOSE(): String {
            return " "
        }
    }

    // Having Statement
    private class HavingStatement constructor(word: String = "\nHAVING ") : WhereStatement(word)

    // Order By Statement
    private class OrderByStatement : BaseStatement("\nORDER BY ", ", ") {
        override fun OPEN(): String {
            return ""
        }

        override fun CLOSE(): String {
            return " "
        }
    }

    private enum class StatementType {
        INSERT, REPLACE, DELETE, UPDATE, SELECT, INSERT_UPDATE
    }

    class SQLWhere(private val sqlBuilder: SQLBuilder) : Serializable {
        private val WHERE = WhereStatement()
        fun AND(): SQLWhere {
            WHERE.AND()
            return this
        }

        fun OR(): SQLWhere {
            WHERE.OR()
            return this
        }

        fun WHERE(wheres: String): SQLWhere {
            WHERE.ADD_VALUES(wheres)
            return this
        }

        fun WHERE_IN(column: String, args: Array<Any?>): SQLWhere {
            args.takeIf { it.isNotEmpty() }?.let {
                WHERE("$column IN(${args.joinToString(", ") { "?" }})")
                sqlBuilder.ARGS(*args)
            }
            return this
        }

    }

    class SQLHaving(private val sqlBuilder: SQLBuilder) : Serializable {
        private val HAVING: HavingStatement = HavingStatement()
        fun AND(): SQLHaving {
            HAVING.AND()
            return this
        }

        fun OR(): SQLHaving {
            HAVING.OR()
            return this
        }

        fun HAVING(having: String): SQLHaving {
            HAVING.ADD_VALUES(having)
            return this
        }

        fun WHERE_IN(column: String, args: Array<Any?>): SQLHaving {
            args.takeIf { it.isNotEmpty() }?.let {
                HAVING("$column IN(${args.joinToString(", ") { "?" }})")
                sqlBuilder.ARGS(*args)
            }
            return this
        }
    }
}